src/mscorlib/shared/System/IO/FileStreamCompletionSource.Win32.cs

```diff
@@ -35,13 +35,33 @@ private unsafe class FileStreamCompletionSource : TaskCompletionSource<int>
             private NativeOverlapped* _overlapped; // Overlapped class responsible for operations in progress when an appdomain unload occurs
             private long _result; // Using long since this needs to be used in Interlocked APIs
 
+            private MemoryHandle _handle; // mutable struct; do not make this readonly
+
             // Using RunContinuationsAsynchronously for compat reasons (old API used Task.Factory.StartNew for continuations)
-            internal FileStreamCompletionSource(FileStream stream, int numBufferedBytes, byte[] bytes)
+            internal FileStreamCompletionSource(FileStream stream, int numBufferedBytes, ReadOnlyMemory<byte> memory)
                 : base(TaskCreationOptions.RunContinuationsAsynchronously)
             {
                 _numBufferedBytes = numBufferedBytes;
                 _stream = stream;
                 _result = NoResult;
+                byte[] bytes = null;
+
+                if (MemoryMarshal.TryGetArray(memory, out ArraySegment<byte> buffer))
+                {
+                    bytes = buffer.Array;
+
+                    // If the memory passed in isn't the stream's internal buffer, then delegate to
+                    // memory.Retain instead of passing the buffer to AllocateNativeOverlapped. In 
+                    // some cases, this will result in less pinning (in case the underlying memory is backed)
+                    // by pre-pinned buffers
```

Comment:
Nit: closing paren should be at the end of the sentence rather than after "backed"